File: csharp/refactoring-methods/05-conditional-expressions.cs
using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// 34. Decomposition of a conditional operator (Decompose Conditional)
///
/// BEFORE: Complex conditional logic
/// </summary>
public class PaymentProcessorBefore
{
    public double CalculateFee(double amount, bool isInternational, bool isPremium)
    {
        double fee = 0;
        if (amount > 100 && isInternational && isPremium)
        {
            fee = amount * 0.05 + 10;
        }
        else if (amount > 100 && isInternational && !isPremium)
        {
            fee = amount * 0.05 + 15;
        }
        else if (amount <= 100 && isInternational)
        {
            fee = amount * 0.03 + 5;
        }
        else
        {
            fee = amount * 0.02;
        }
        return fee;
    }
}

/// <summary>
/// AFTER: Decompose conditional
/// </summary>
public class PaymentProcessorAfter
{
    public double CalculateFee(double amount, bool isInternational, bool isPremium)
    {
        if (IsHighValueInternationalPremium(amount, isInternational, isPremium))
        {
            return CalculateHighValueInternationalPremiumFee(amount);
        }
        else if (IsHighValueInternationalStandard(amount, isInternational, isPremium))
        {
            return CalculateHighValueInternationalStandardFee(amount);
        }
        else if (IsLowValueInternational(amount, isInternational))
        {
            return CalculateLowValueInternationalFee(amount);
        }
        else
        {
            return CalculateDomesticFee(amount);
        }
    }

    private bool IsHighValueInternationalPremium(double amount, bool isInternational, bool isPremium)
    {
        return amount > 100 && isInternational && isPremium;
    }

    private bool IsHighValueInternationalStandard(double amount, bool isInternational, bool isPremium)
    {
        return amount > 100 && isInternational && !isPremium;
    }

    private bool IsLowValueInternational(double amount, bool isInternational)
    {
        return amount <= 100 && isInternational;
    }

    private double CalculateHighValueInternationalPremiumFee(double amount)
    {
        return amount * 0.05 + 10;
    }

    private double CalculateHighValueInternationalStandardFee(double amount)
    {
        return amount * 0.05 + 15;
    }

    private double CalculateLowValueInternationalFee(double amount)
    {
        return amount * 0.03 + 5;
    }

    private double CalculateDomesticFee(double amount)
    {
        return amount * 0.02;
    }
}

/// <summary>
/// 35. Consolidation of a conditional expression (Consolidate Conditional Expression)
///
/// BEFORE: Multiple conditionals with same result
/// </summary>
public class InsuranceCalculatorBefore
{
    public bool IsEligibleForDiscount(int age, bool isStudent, bool hasGoodRecord)
    {
        if (age < 25) return false;
        if (isStudent) return true;
        if (hasGoodRecord) return true;
        return false;
    }
}

/// <summary>
/// AFTER: Consolidate conditionals
/// </summary>
public class InsuranceCalculatorAfter
{
    public bool IsEligibleForDiscount(int age, bool isStudent, bool hasGoodRecord)
    {
        return age >= 25 && (isStudent || hasGoodRecord);
    }
}

/// <summary>
/// 36. Consolidation of duplicate conditional fragments
/// (Consolidate Duplicate Conditional Fragments)
///
/// BEFORE: Duplicate code in conditional branches
/// </summary>
public class FileProcessorBefore
{
    public void ProcessFile(object file)
    {
        if (IsValidFile(file))
        {
            LogProcessing(file);
            ValidateContent(file);
            SaveToDatabase(file);
            SendNotification(file);
        }
        else
        {
            LogError(file);
            SendNotification(file); // Duplicate
        }
    }

    private void SendNotification(object file)
    {
        // Send notification logic
    }

    private bool IsValidFile(object file) => true;
    private void LogProcessing(object file) { }
    private void ValidateContent(object file) { }
    private void SaveToDatabase(object file) { }
    private void LogError(object file) { }
}

/// <summary>
/// AFTER: Consolidate duplicate fragments
/// </summary>
public class FileProcessorAfter
{
    public void ProcessFile(object file)
    {
        SendNotification(file); // Moved outside conditional

        if (IsValidFile(file))
        {
            LogProcessing(file);
            ValidateContent(file);
            SaveToDatabase(file);
        }
        else
        {
            LogError(file);
        }
    }

    private void SendNotification(object file)
    {
        // Send notification logic
    }

    private bool IsValidFile(object file) => true;
    private void LogProcessing(object file) { }
    private void ValidateContent(object file) { }
    private void SaveToDatabase(object file) { }
    private void LogError(object file) { }
}

/// <summary>
/// 37. Remove Control Flag
///
/// BEFORE: Control flag to break out of loop
/// </summary>
public class DataProcessorBefore
{
    public object FindPerson(List<Dictionary<string, object>> people, string name)
    {
        object found = null;
        foreach (var person in people)
        {
            if ((string)person["name"] == name)
            {
                found = person;
                break; // Control flag usage
            }
        }
        return found;
    }
}

/// <summary>
/// AFTER: Remove control flag
/// </summary>
public class DataProcessorAfter
{
    public Dictionary<string, object> FindPerson(List<Dictionary<string, object>> people, string name)
    {
        foreach (var person in people)
        {
            if ((string)person["name"] == name)
            {
                return person; // Direct return
            }
        }
        return null;
    }
}

/// <summary>
/// 38. Replacing Nested Conditional statements with a boundary operator
/// (Replace Nested Conditional with Guard Clauses)
///
/// BEFORE: Nested conditionals
/// </summary>
public class PaymentValidatorBefore
{
    public bool IsValidPayment(Dictionary<string, object> payment)
    {
        if ((double)payment["amount"] > 0)
        {
            if (payment["cardNumber"] != null)
            {
                if (((string)payment["cardNumber"]).Length == 16)
                {
                    if (IsValidExpiry((string)payment["expiry"]))
                    {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private bool IsValidExpiry(string expiry)
    {
        return DateTime.Parse(expiry) > DateTime.Now;
    }
}

/// <summary>
/// AFTER: Replace with guard clauses
/// </summary>
public class PaymentValidatorAfter
{
    public bool IsValidPayment(Dictionary<string, object> payment)
    {
        if ((double)payment["amount"] <= 0)
        {
            return false;
        }

        if (payment["cardNumber"] == null)
        {
            return false;
        }

        if (((string)payment["cardNumber"]).Length != 16)
        {
            return false;
        }

        if (!IsValidExpiry((string)payment["expiry"]))
        {
            return false;
        }

        return true;
    }

    private bool IsValidExpiry(string expiry)
    {
        return DateTime.Parse(expiry) > DateTime.Now;
    }
}

/// <summary>
/// 39. Replacing a conditional operator with polymorphism (Replace Conditional with Polymorphism)
///
/// BEFORE: Type checking with conditionals
/// </summary>
public class BirdBefore
{
    public const string EUROPEAN = "european";
    public const string AFRICAN = "african";
    public const string NORWEGIAN_BLUE = "norwegian_blue";

    private string type;
    private int voltage;
    private bool isNailed;

    public BirdBefore(string type)
    {
        this.type = type;
    }

    public double GetSpeed()
    {
        switch (type)
        {
            case EUROPEAN:
                return GetBaseSpeed();
            case AFRICAN:
                return GetBaseSpeed() - voltage * 2;
            case NORWEGIAN_BLUE:
                return isNailed ? 0 : GetBaseSpeed();
            default:
                return GetBaseSpeed();
        }
    }

    private double GetBaseSpeed()
    {
        return 10;
    }
}

/// <summary>
/// AFTER: Replace conditional with polymorphism
/// </summary>
public abstract class BirdAfter
{
    public abstract double GetSpeed();

    protected double GetBaseSpeed()
    {
        return 10;
    }
}

public class EuropeanSwallow : BirdAfter
{
    public override double GetSpeed()
    {
        return GetBaseSpeed();
    }
}

public class AfricanSwallow : BirdAfter
{
    private int voltage;

    public AfricanSwallow(int voltage)
    {
        this.voltage = voltage;
    }

    public override double GetSpeed()
    {
        return GetBaseSpeed() - voltage * 2;
    }
}

public class NorwegianBlueParrot : BirdAfter
{
    private bool isNailed;

    public NorwegianBlueParrot(bool isNailed)
    {
        this.isNailed = isNailed;
    }

    public override double GetSpeed()
    {
        return isNailed ? 0 : GetBaseSpeed();
    }
}

/// <summary>
/// 40. Introduction of the object (Introduce Object)
///
/// BEFORE: Primitive obsession with conditionals
/// </summary>
public class UserValidatorBefore
{
    public object ValidateUser(Dictionary<string, object> user)
    {
        if (string.IsNullOrEmpty((string)user["name"]))
        {
            return "Name is required";
        }

        if (((string)user["name"]).Length < 2)
        {
            return "Name must be at least 2 characters";
        }

        if (!IsValidEmail((string)user["email"]))
        {
            return "Invalid email format";
        }

        return true;
    }

    private bool IsValidEmail(string email)
    {
        try
        {
            var addr = new System.Net.Mail.MailAddress(email);
            return addr.Address == email;
        }
        catch
        {
            return false;
        }
    }
}

/// <summary>
/// AFTER: Introduce validation result object
/// </summary>
public class ValidationResult
{
    private bool isValid;
    private List<string> errors;

    public ValidationResult(bool isValid = true, List<string> errors = null)
    {
        this.isValid = isValid;
        this.errors = errors ?? new List<string>();
    }

    public bool IsValid()
    {
        return isValid;
    }

    public List<string> GetErrors()
    {
        return new List<string>(errors);
    }

    public ValidationResult AddError(string error)
    {
        isValid = false;
        errors.Add(error);
        return this;
    }
}

public class UserValidatorAfter
{
    public ValidationResult ValidateUser(Dictionary<string, object> user)
    {
        var result = new ValidationResult();

        if (string.IsNullOrEmpty((string)user["name"]))
        {
            result.AddError("Name is required");
        }

        if (((string)user["name"]).Length < 2)
        {
            result.AddError("Name must be at least 2 characters");
        }

        if (!IsValidEmail((string)user["email"]))
        {
            result.AddError("Invalid email format");
        }

        return result;
    }

    private bool IsValidEmail(string email)
    {
        try
        {
            var addr = new System.Net.Mail.MailAddress(email);
            return addr.Address == email;
        }
        catch
        {
            return false;
        }
    }
}

/// <summary>
/// 41. Introduction of the statement (Introduction Statement)
///
/// BEFORE: Magic assertion
/// </summary>
public class AccountAssertion
{
    private double balance;

    public void Withdraw(double amount)
    {
        System.Diagnostics.Debug.Assert(amount > 0 && amount <= balance);
        balance -= amount;
    }
}

/// <summary>
/// AFTER: Introduce assertion method
/// </summary>
public class AccountAssertionAfter
{
    private double balance;

    public void Withdraw(double amount)
    {
        AssertValidWithdrawal(amount);
        balance -= amount;
    }

    private void AssertValidWithdrawal(double amount)
    {
        System.Diagnostics.Debug.Assert(amount > 0 && amount <= balance, "Invalid withdrawal amount");
    }
}

File: python/refactoring-methods/05-conditional-expressions.py
import re
from abc import ABC, abstractmethod
from datetime import datetime

# 34. Decomposition of a conditional operator (Decompose Conditional)
#
# BEFORE: Complex conditional logic
class PaymentProcessorBefore:
    def calculate_fee(self, amount, is_international, is_premium):
        fee = 0
        if amount > 100 and is_international and is_premium:
            fee = amount * 0.05 + 10
        elif amount > 100 and is_international and not is_premium:
            fee = amount * 0.05 + 15
        elif amount <= 100 and is_international:
            fee = amount * 0.03 + 5
        else:
            fee = amount * 0.02
        return fee

# AFTER: Decompose conditional
class PaymentProcessorAfter:
    def calculate_fee(self, amount, is_international, is_premium):
        if self._is_high_value_international_premium(amount, is_international, is_premium):
            return self._calculate_high_value_international_premium_fee(amount)
        elif self._is_high_value_international_standard(amount, is_international, is_premium):
            return self._calculate_high_value_international_standard_fee(amount)
        elif self._is_low_value_international(amount, is_international):
            return self._calculate_low_value_international_fee(amount)
        else:
            return self._calculate_domestic_fee(amount)

    def _is_high_value_international_premium(self, amount, is_international, is_premium):
        return amount > 100 and is_international and is_premium

    def _is_high_value_international_standard(self, amount, is_international, is_premium):
        return amount > 100 and is_international and not is_premium

    def _is_low_value_international(self, amount, is_international):
        return amount <= 100 and is_international

    def _calculate_high_value_international_premium_fee(self, amount):
        return amount * 0.05 + 10

    def _calculate_high_value_international_standard_fee(self, amount):
        return amount * 0.05 + 15

    def _calculate_low_value_international_fee(self, amount):
        return amount * 0.03 + 5

    def _calculate_domestic_fee(self, amount):
        return amount * 0.02

# 35. Consolidation of a conditional expression (Consolidate Conditional Expression)
#
# BEFORE: Multiple conditionals with same result
class InsuranceCalculatorBefore:
    def is_eligible_for_discount(self, age, is_student, has_good_record):
        if age < 25:
            return False
        if is_student:
            return True
        if has_good_record:
            return True
        return False

# AFTER: Consolidate conditionals
class InsuranceCalculatorAfter:
    def is_eligible_for_discount(self, age, is_student, has_good_record):
        return age >= 25 and (is_student or has_good_record)

# 36. Consolidation of duplicate conditional fragments
# (Consolidate Duplicate Conditional Fragments)
#
# BEFORE: Duplicate code in conditional branches
class FileProcessorBefore:
    def process_file(self, file_path):
        if self._is_valid_file(file_path):
            self._log_processing(file_path)
            self._validate_content(file_path)
            self._save_to_database(file_path)
            self._send_notification(file_path)
        else:
            self._log_error(file_path)
            self._send_notification(file_path)  # Duplicate

    def _send_notification(self, file_path):
        # Send notification logic
        pass

# AFTER: Consolidate duplicate fragments
class FileProcessorAfter:
    def process_file(self, file_path):
        self._send_notification(file_path)  # Moved outside conditional

        if self._is_valid_file(file_path):
            self._log_processing(file_path)
            self._validate_content(file_path)
            self._save_to_database(file_path)
        else:
            self._log_error(file_path)

    def _send_notification(self, file_path):
        # Send notification logic
        pass

# 37. Remove Control Flag
#
# BEFORE: Control flag to break out of loop
class DataProcessorBefore:
    def find_person(self, people, name):
        found = False
        for person in people:
            if person['name'] == name:
                found = person
                break  # Control flag usage
        return found

# AFTER: Remove control flag
class DataProcessorAfter:
    def find_person(self, people, name):
        for person in people:
            if person['name'] == name:
                return person  # Direct return
        return False

# 38. Replacing Nested Conditional statements with a boundary operator
# (Replace Nested Conditional with Guard Clauses)
#
# BEFORE: Nested conditionals
class PaymentValidatorBefore:
    def is_valid_payment(self, payment):
        if payment['amount'] > 0:
            if payment['card_number'] is not None:
                if len(payment['card_number']) == 16:
                    if self._is_valid_expiry(payment['expiry']):
                        return True
        return False

    def _is_valid_expiry(self, expiry):
        return datetime.strptime(expiry, '%Y-%m-%d') > datetime.now()

# AFTER: Replace with guard clauses
class PaymentValidatorAfter:
    def is_valid_payment(self, payment):
        if payment['amount'] <= 0:
            return False

        if payment['card_number'] is None:
            return False

        if len(payment['card_number']) != 16:
            return False

        if not self._is_valid_expiry(payment['expiry']):
            return False

        return True

    def _is_valid_expiry(self, expiry):
        return datetime.strptime(expiry, '%Y-%m-%d') > datetime.now()

# 39. Replacing a conditional operator with polymorphism (Replace Conditional with Polymorphism)
#
# BEFORE: Type checking with conditionals
class BirdBefore:
    EUROPEAN = 'european'
    AFRICAN = 'african'
    NORWEGIAN_BLUE = 'norwegian_blue'

    def __init__(self, bird_type):
        self._type = bird_type

    def get_speed(self):
        if self._type == self.EUROPEAN:
            return self._get_base_speed()
        elif self._type == self.AFRICAN:
            return self._get_base_speed() - self._voltage * 2
        elif self._type == self.NORWEGIAN_BLUE:
            return 0 if self._is_nailed else self._get_base_speed()
        else:
            return self._get_base_speed()

    def _get_base_speed(self):
        return 10

# AFTER: Replace conditional with polymorphism
class BirdAfter(ABC):
    @abstractmethod
    def get_speed(self):
        pass

    def _get_base_speed(self):
        return 10

class EuropeanSwallow(BirdAfter):
    def get_speed(self):
        return self._get_base_speed()

class AfricanSwallow(BirdAfter):
    def __init__(self, voltage):
        self._voltage = voltage

    def get_speed(self):
        return self._get_base_speed() - self._voltage * 2

class NorwegianBlueParrot(BirdAfter):
    def __init__(self, is_nailed):
        self._is_nailed = is_nailed

    def get_speed(self):
        return 0 if self._is_nailed else self._get_base_speed()

# 40. Introduction of the object (Introduce Object)
#
# BEFORE: Primitive obsession with conditionals
class UserValidatorBefore:
    def validate_user(self, user):
        if not user['name']:
            return 'Name is required'

        if len(user['name']) < 2:
            return 'Name must be at least 2 characters'

        if not self._is_valid_email(user['email']):
            return 'Invalid email format'

        return True

    def _is_valid_email(self, email):
        # Simple email validation
        return '@' in email and '.' in email

# AFTER: Introduce validation result object
class ValidationResult:
    def __init__(self, is_valid=True, errors=None):
        self._is_valid = is_valid
        self._errors = errors or []

    def is_valid(self):
        return self._is_valid

    def get_errors(self):
        return self._errors

    def add_error(self, error):
        self._is_valid = False
        self._errors.append(error)
        return self

class UserValidatorAfter:
    def validate_user(self, user):
        result = ValidationResult()

        if not user['name']:
            result.add_error('Name is required')

        if len(user['name']) < 2:
            result.add_error('Name must be at least 2 characters')

        if not self._is_valid_email(user['email']):
            result.add_error('Invalid email format')

        return result

    def _is_valid_email(self, email):
        # Simple email validation
        return '@' in email and '.' in email

# 41. Introduction of the statement (Introduction Statement)
#
# BEFORE: Magic assertion
class AccountAssertion:
    def __init__(self, balance):
        self._balance = balance

    def withdraw(self, amount):
        assert amount > 0 and amount <= self._balance
        self._balance -= amount

# AFTER: Introduce assertion method
class AccountAssertionAfter:
    def __init__(self, balance):
        self._balance = balance

    def withdraw(self, amount):
        self._assert_valid_withdrawal(amount)
        self._balance -= amount

    def _assert_valid_withdrawal(self, amount):
        assert amount > 0 and amount <= self._balance, 'Invalid withdrawal amount'
